crates/subspace-networking/src/node_runner.rs

```diff
@@ -360,6 +388,22 @@ where
             .get_closest_peers(random_peer_id);
     }
 
+    fn handle_removed_address_event(&mut self, event: PeerAddressRemovedEvent) {
+        trace!(?event, "Peer addressed removed event.",);
+
+        if event.last_address {
+            self.swarm
+                .behaviour_mut()
+                .kademlia
+                .remove_peer(&event.peer_id);
+        } else {
+            self.swarm
+                .behaviour_mut()
+                .kademlia
+                .remove_address(&event.peer_id, &event.address);
+        }
```

Comment:
Yes, it does. However, I think there is a bug here anyway. The problem will arise when we remove known addresses that are unresponsive and remove other addresses (via `remove_peer`) that we don't know about because we didn't dial them previously. I changed the logic to handle the `UnroutablePeer` event instead.

Reply:
Okay, that makes sense and I think we should then replace `handle_removed_address_event` with `handle_removed_peer`, but even then it seems problematic that we may not remember all the peers in Kademlia anyway, thus unable to remove them. Should known peers be unbounded in memory then and bounded when written to disk (by unbounded I mean bounded by Kademlia externally)?

Reply:
What I meant is that our known peers cache is limited in size. It seems possible that we will have some peers added to Kademlia, but not in known peers. If that is the case, they will never be removed from Kademlia and we do not have event about things evicted from LRU cache of known peers.

Reply:
Can we derive cache limits in your pending PR from the Kademlia internals?

Reply:
Not dynamically, but if we have fixed Kademlia parameters and we know upper bound then yes. I agree this PR is an improvement, I'm just trying to understand whether it is a 100% solution to the problem.